# 1. Importing libraries

In [25]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [26]:
# Turning your project folder path into a string as prepared data.
path1 = r'/Users/elizabetenakamura/Documents/projects/python/data/prepared'

# Importing Data Frames

In [27]:
# Import and identify dataframes
ords_prods_merge = pd.read_pickle(os.path.join(path1,'ords_prods_merged_derived.pkl'))

In [28]:
# Create a subset of 1 million rows.
df = ords_prods_merge[:1000000]

In [29]:
df.shape

(1000000, 24)

In [30]:
df.head(10)

Unnamed: 0.1_x  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0               0             0   2539329        1    prior             1   
1               0             0   2539329        1    prior             1   
2               0             0   2539329        1    prior             1   
3               0             0   2539329        1    prior             1   
4               0             0   2539329        1    prior             1   
5               1             1   2398795        1    prior             2   
6               1             1   2398795        1    prior             2   
7               1             1   2398795        1    prior             2   
8               1             1   2398795        1    prior             2   
9               1             1   2398795        1    prior             2   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   2                  8                     NaN   
1                   2                  8                     NaN   
2                   2                  8                     NaN   
3                   2                  8                     NaN   
4                   2                  8                     NaN   
5                   3                  7                    15.0   
6                   3                  7                    15.0   
7                   3                  7                    15.0   
8                   3                  7                    15.0   
9                   3                  7                    15.0   

   new_customer  ...  Unnamed: 0.1_y  Unnamed: 0_y  \
0          True  ...             192           195   
1          True  ...           14069         14084   
2          True  ...           12412         12427   
3          True  ...           26073         26089   
4          True  ...           26390         26406   
5         False  ...             192           195   
6         False  ...           10243         10258   
7         False  ...           12412         12427   
8         False  ...           13161         13176   
9         False  ...           26073         26089   

                              product_name aisle_id  department_id  prices  \
0                                     Soda       77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk       91             16    12.5   
2                      Original Beef Jerky       23             19     4.4   
3               Aged White Cheddar Popcorn       23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls       54             17     1.0   
5                                     Soda       77              7     9.0   
6                               Pistachios      117             19     3.0   
7                      Original Beef Jerky       23             19     4.4   
8                   Bag of Organic Bananas       24              4    10.3   
9               Aged White Cheddar Popcorn       23             19     4.7   

  match  busiest_days  busiest_period_of_day  busiest_period_of_days  
0  both   Normal days         Average orders          Average orders  
1  both   Normal days         Average orders          Average orders  
2  both   Normal days         Average orders          Average orders  
3  both   Normal days         Average orders          Average orders  
4  both   Normal days         Average orders          Average orders  
5  both   Normal days         Average orders          Average orders  
6  both   Normal days         Average orders          Average orders  
7  both   Normal days         Average orders          Average orders  
8  both   Normal days         Average orders          Average orders  
9  both   Normal days         Average orders          Average orders  

[10 rows x 24 columns]

# 2. Grouping Data with Pandas

In [31]:
# Grouping data with pandas.
ords_prods_merge.groupby('product_name')

# 3. Aggregating Data

In [32]:
# Subset aggregation.
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

In [33]:
# Entire dataframe aggregation.
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

# 3. The overall trends in the mean, min, and max values might change when you analyze subsets. 
Overall Trends:
The entire dataframe generally shows higher mean order numbers across departments, suggesting that it encompasses a broader spectrum of customer behaviors compared to the subset.

Notable Differences:
Department 21 exhibits the highest mean in both datasets, with a higher value in the subset (25.53) compared to the entire dataframe (22.90). Additionally, Department 10 shows a significant increase in the entire dataframe (20.19) compared to the subset (18.68).

Consistent Departments:
Some departments, such as 13 and 17, exhibit minimal differences between the two datasets. However, other departments, including 3, 9, and 12, display more pronounced variations.

Conclusion:
Overall, the entire dataframe reveals more frequent and diverse ordering patterns across departments when compared to the subset.

# 4. Aggregating Data with Transform

In [34]:
 # Create max_order column.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/95/tcwc05191szglpgs3jtdf6tm0000gn/T/ipykernel_9521/2964966201.py:2: FutureWarning: The provided callable <function max at 0x106bbb560> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [35]:
ords_prods_merge.drop(columns = ['eval_set'], inplace = True)

In [36]:
ords_prods_merge.head(15)

Unnamed: 0.1_x  Unnamed: 0_x  order_id  user_id  order_number  \
0                0             0   2539329        1             1   
1                0             0   2539329        1             1   
2                0             0   2539329        1             1   
3                0             0   2539329        1             1   
4                0             0   2539329        1             1   
5                1             1   2398795        1             2   
6                1             1   2398795        1             2   
7                1             1   2398795        1             2   
8                1             1   2398795        1             2   
9                1             1   2398795        1             2   
10               1             1   2398795        1             2   
11               2             2    473747        1             3   
12               2             2    473747        1             3   
13               2             2    473747        1             3   
14               2             2    473747        1             3   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   
5                    3                  7                    15.0   
6                    3                  7                    15.0   
7                    3                  7                    15.0   
8                    3                  7                    15.0   
9                    3                  7                    15.0   
10                   3                  7                    15.0   
11                   3                 12                    21.0   
12                   3                 12                    21.0   
13                   3                 12                    21.0   
14                   3                 12                    21.0   

    new_customer  product_id  ...  Unnamed: 0_y  \
0           True         196  ...           195   
1           True       14084  ...         14084   
2           True       12427  ...         12427   
3           True       26088  ...         26089   
4           True       26405  ...         26406   
5          False         196  ...           195   
6          False       10258  ...         10258   
7          False       12427  ...         12427   
8          False       13176  ...         13176   
9          False       26088  ...         26089   
10         False       13032  ...         13032   
11         False         196  ...           195   
12         False       12427  ...         12427   
13         False       10258  ...         10258   
14         False       25133  ...         25134   

                               product_name aisle_id  department_id  prices  \
0                                      Soda       77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk       91             16    12.5   
2                       Original Beef Jerky       23             19     4.4   
3                Aged White Cheddar Popcorn       23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls       54             17     1.0   
5                                      Soda       77              7     9.0   
6                                Pistachios      117             19     3.0   
7                       Original Beef Jerky       23             19     4.4   
8                    Bag of Organic Bananas       24              4    10.3   
9                Aged White Cheddar Popcorn       23             19     4.7   
10                    Cinnamon Toast Crunch      121             14     4.0   
11                           

In [37]:
ords_prods_merge.head(100)

Unnamed: 0.1_x  Unnamed: 0_x  order_id  user_id  order_number  \
0                0             0   2539329        1             1   
1                0             0   2539329        1             1   
2                0             0   2539329        1             1   
3                0             0   2539329        1             1   
4                0             0   2539329        1             1   
5                1             1   2398795        1             2   
6                1             1   2398795        1             2   
7                1             1   2398795        1             2   
8                1             1   2398795        1             2   
9                1             1   2398795        1             2   
10               1             1   2398795        1             2   
11               2             2    473747        1             3   
12               2             2    473747        1             3   
13               2             2    473747        1             3   
14               2             2    473747        1             3   
15               2             2    473747        1             3   
16               3             3   2254736        1             4   
17               3             3   2254736        1             4   
18               3             3   2254736        1             4   
19               3             3   2254736        1             4   
20               3             3   2254736        1             4   
21               4             4    431534        1             5   
22               4             4    431534        1             5   
23               4             4    431534        1             5   
24               4             4    431534        1             5   
25               4             4    431534        1             5   
26               4             4    431534        1             5   
27               4             4    431534        1             5   
28               4             4    431534        1             5   
29               5             5   3367565        1             6   
30               5             5   3367565        1             6   
31               5             5   3367565        1             6   
32               5             5   3367565        1             6   
33               6             6    550135        1             7   
34               6             6    550135        1             7   
35               6             6    550135        1             7   
36               6             6    550135        1             7   
37               6             6    550135        1             7   
38               7             7   3108588        1             8   
39               7             7   3108588        1             8   
40               7             7   3108588        1             8   
41               7             7   3108588        1             8   
42               7             7   3108588        1             8   
43               7             7   3108588        1             8   
44               8             8   2295261        1             9   
45               8             8   2295261        1             9   
46               8             8   2295261        1             9   
47               8             8   2295261        1             9   
48               8             8   2295261        1             9   
49               8             8   2295261        1             9   
50               9             9   2550362        1            10   
51               9             9   2550362        1            10   
52               9             9   2550362        1            10   
53               9             9   2550362        1            10   
54               9             9   2550362        1            10   
55               9             9   2550362        1            10   
56               9             9   2550362        1            10   
57

In [38]:
pd.options.display.max_rows = None

# 5. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [42]:
# Create a loyalt_flay using loc function.
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [43]:
# Frequency of each distinct row in the Dataframe.
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [45]:
# Checking the Updated Output.
ords_prods_merge.head(15)

Unnamed: 0.1_x  Unnamed: 0_x  order_id  user_id  order_number  \
0                0             0   2539329        1             1   
1                0             0   2539329        1             1   
2                0             0   2539329        1             1   
3                0             0   2539329        1             1   
4                0             0   2539329        1             1   
5                1             1   2398795        1             2   
6                1             1   2398795        1             2   
7                1             1   2398795        1             2   
8                1             1   2398795        1             2   
9                1             1   2398795        1             2   
10               1             1   2398795        1             2   
11               2             2    473747        1             3   
12               2             2    473747        1             3   
13               2             2    473747        1             3   
14               2             2    473747        1             3   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   
5                    3                  7                    15.0   
6                    3                  7                    15.0   
7                    3                  7                    15.0   
8                    3                  7                    15.0   
9                    3                  7                    15.0   
10                   3                  7                    15.0   
11                   3                 12                    21.0   
12                   3                 12                    21.0   
13                   3                 12                    21.0   
14                   3                 12                    21.0   

    new_customer  product_id  ...                             product_name  \
0           True         196  ...                                     Soda   
1           True       14084  ...  Organic Unsweetened Vanilla Almond Milk   
2           True       12427  ...                      Original Beef Jerky   
3           True       26088  ...               Aged White Cheddar Popcorn   
4           True       26405  ...         XL Pick-A-Size Paper Towel Rolls   
5          False         196  ...                                     Soda   
6          False       10258  ...                               Pistachios   
7          False       12427  ...                      Original Beef Jerky   
8          False       13176  ...                   Bag of Organic Bananas   
9          False       26088  ...               Aged White Cheddar Popcorn   
10         False       13032  ...                    Cinnamon Toast Crunch   
11         False         196  ...                                     Soda   
12         False       12427  ...                      Original Beef Jerky   
13         False       10258  ...                               Pistachios   
14         False       25133  ...                    Organic String Cheese   

    aisle_id department_id  prices  match busiest_days  busiest_period_of_day  \
0         77             7     9.0   both  Normal days         Average orders   
1         91            16    12.5   both  Normal days         Average orders   
2         23            19     4.4   both  Normal days         Average orders   
3         23            19     4.7   both  Normal days         Average orders   
4         54            17     1.0   both  Normal days         Average orders   
5         77             7     9.0   both  Normal days     

# Analysis of Spending Habits by Loyalty Category

In [47]:
# Basic Statistics of Spending Habits by Loyalty Category.
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer     9.579346  1.0  14900.0
New customer      10.059889  1.0  14900.0
Regular customer   9.894698  1.0  14900.0

# 6. Create a spending flag for each user based on the average price across all their orders.

In [49]:
# Create a new column for the average spending of a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [50]:
# Check the new column
ords_prods_merge.head(20)

Unnamed: 0.1_x  Unnamed: 0_x  order_id  user_id  order_number  \
0                0             0   2539329        1             1   
1                0             0   2539329        1             1   
2                0             0   2539329        1             1   
3                0             0   2539329        1             1   
4                0             0   2539329        1             1   
5                1             1   2398795        1             2   
6                1             1   2398795        1             2   
7                1             1   2398795        1             2   
8                1             1   2398795        1             2   
9                1             1   2398795        1             2   
10               1             1   2398795        1             2   
11               2             2    473747        1             3   
12               2             2    473747        1             3   
13               2             2    473747        1             3   
14               2             2    473747        1             3   
15               2             2    473747        1             3   
16               3             3   2254736        1             4   
17               3             3   2254736        1             4   
18               3             3   2254736        1             4   
19               3             3   2254736        1             4   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   
5                    3                  7                    15.0   
6                    3                  7                    15.0   
7                    3                  7                    15.0   
8                    3                  7                    15.0   
9                    3                  7                    15.0   
10                   3                  7                    15.0   
11                   3                 12                    21.0   
12                   3                 12                    21.0   
13                   3                 12                    21.0   
14                   3                 12                    21.0   
15                   3                 12                    21.0   
16                   4                  7                    29.0   
17                   4                  7                    29.0   
18                   4                  7                    29.0   
19                   4                  7                    29.0   

    new_customer  product_id  ...  aisle_id  department_id prices  match  \
0           True         196  ...        77              7    9.0   both   
1           True       14084  ...        91             16   12.5   both   
2           True       12427  ...        23             19    4.4   both   
3           True       26088  ...        23             19    4.7   both   
4           True       26405  ...        54             17    1.0   both   
5          False         196  ...        77              7    9.0   both   
6          False       10258  ...       117             19    3.0   both   
7          False       12427  ...        23             19    4.4   both   
8          False       13176  ...        24              4   10.3   both   
9          False       26088  ...        23             19    4.7   both   
10         False       13032  ...       121             14    4.0   both   
11         False         196  ...        77              7    9.0   both   
12         False       12427  ...        23             19    4.4   both   
13         False       10258  ...       

In [51]:
# Assign spender tag to each customer
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='low_spender'

In [52]:
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'high_spender'

In [53]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
low_spender     31828060
high_spender      576799
Name: count, dtype: int64

# 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [55]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [56]:
ords_prods_merge.head(20)

Unnamed: 0.1_x  Unnamed: 0_x  order_id  user_id  order_number  \
0                0             0   2539329        1             1   
1                0             0   2539329        1             1   
2                0             0   2539329        1             1   
3                0             0   2539329        1             1   
4                0             0   2539329        1             1   
5                1             1   2398795        1             2   
6                1             1   2398795        1             2   
7                1             1   2398795        1             2   
8                1             1   2398795        1             2   
9                1             1   2398795        1             2   
10               1             1   2398795        1             2   
11               2             2    473747        1             3   
12               2             2    473747        1             3   
13               2             2    473747        1             3   
14               2             2    473747        1             3   
15               2             2    473747        1             3   
16               3             3   2254736        1             4   
17               3             3   2254736        1             4   
18               3             3   2254736        1             4   
19               3             3   2254736        1             4   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   
5                    3                  7                    15.0   
6                    3                  7                    15.0   
7                    3                  7                    15.0   
8                    3                  7                    15.0   
9                    3                  7                    15.0   
10                   3                  7                    15.0   
11                   3                 12                    21.0   
12                   3                 12                    21.0   
13                   3                 12                    21.0   
14                   3                 12                    21.0   
15                   3                 12                    21.0   
16                   4                  7                    29.0   
17                   4                  7                    29.0   
18                   4                  7                    29.0   
19                   4                  7                    29.0   

    new_customer  product_id  ...  prices  match  busiest_days  \
0           True         196  ...     9.0   both   Normal days   
1           True       14084  ...    12.5   both   Normal days   
2           True       12427  ...     4.4   both   Normal days   
3           True       26088  ...     4.7   both   Normal days   
4           True       26405  ...     1.0   both   Normal days   
5          False         196  ...     9.0   both   Normal days   
6          False       10258  ...     3.0   both   Normal days   
7          False       12427  ...     4.4   both   Normal days   
8          False       13176  ...    10.3   both   Normal days   
9          False       26088  ...     4.7   both   Normal days   
10         False       13032  ...     4.0   both   Normal days   
11         False         196  ...     9.0   both   Normal days   
12         False       12427  ...     4.4   both   Normal days   
13         False       10258  ...     3.0   both   Normal days   
14         False       25133  ...     8.6   both   Normal days   
15         False       30450  ...     3.2   both

In [57]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <= 20), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [58]:
# Check value counts for new column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [59]:
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(60)

user_id  customer_frequency         frequency_flag
0         1                20.5  Non-frequent customer
1         1                20.5  Non-frequent customer
2         1                20.5  Non-frequent customer
3         1                20.5  Non-frequent customer
4         1                20.5  Non-frequent customer
5         1                20.5  Non-frequent customer
6         1                20.5  Non-frequent customer
7         1                20.5  Non-frequent customer
8         1                20.5  Non-frequent customer
9         1                20.5  Non-frequent customer
10        1                20.5  Non-frequent customer
11        1                20.5  Non-frequent customer
12        1                20.5  Non-frequent customer
13        1                20.5  Non-frequent customer
14        1                20.5  Non-frequent customer
15        1                20.5  Non-frequent customer
16        1                20.5  Non-frequent customer
17        1                20.5  Non-frequent customer
18        1                20.5  Non-frequent customer
19        1                20.5  Non-frequent customer
20        1                20.5  Non-frequent customer
21        1                20.5  Non-frequent customer
22        1                20.5  Non-frequent customer
23        1                20.5  Non-frequent customer
24        1                20.5  Non-frequent customer
25        1                20.5  Non-frequent customer
26        1                20.5  Non-frequent customer
27        1                20.5  Non-frequent customer
28        1                20.5  Non-frequent customer
29        1                20.5  Non-frequent customer
30        1                20.5  Non-frequent customer
31        1                20.5  Non-frequent customer
32        1                20.5  Non-frequent customer
33        1                20.5  Non-frequent customer
34        1                20.5  Non-frequent customer
35        1                20.5  Non-frequent customer
36        1                20.5  Non-frequent customer
37        1                20.5  Non-frequent customer
38        1                20.5  Non-frequent customer
39        1                20.5  Non-frequent customer
40        1                20.5  Non-frequent customer
41        1                20.5  Non-frequent customer
42        1                20.5  Non-frequent customer
43        1                20.5  Non-frequent customer
44        1                20.5  Non-frequent customer
45        1                20.5  Non-frequent customer
46        1                20.5  Non-frequent customer
47        1                20.5  Non-frequent customer
48        1                20.5  Non-frequent customer
49        1                20.5  Non-frequent customer
50        1                20.5  Non-frequent customer
51        1                20.5  Non-frequent customer
52        1                20.5  Non-frequent customer
53        1                20.5  Non-frequent customer
54        1                20.5  Non-frequent customer
55        1                20.5  Non-frequent customer
56        1                20.5  Non-frequent customer
57        1                20.5  Non-frequent customer
58        1                20.5  Non-frequent customer
59        2                13.0       Regular customer

In [60]:
# Export dataframe with newly derived columns.
ords_prods_merge.to_pickle(os.path.join(path1, 'ords_prods_new_merged_derived.pkl'))